File: programmers/high_score_kit/완전탐색/소수찾기.py
from itertools import permutations
import math


def isPrime(num):
    count = 0
    if num <= 1:
        return False
    for i in range(2, int(math.sqrt(num)) + 1):
        if num % i == 0:
            count += 1
    if count >= 1:
        return False
    return True


def solution(numbers):
    # 문자열을 문자하나하나 나눈 다음, 가능한 조합을 구한다.
    # 구한 조합들 중 소수의 개수를 센다.
    answer = 0
    datas = list()
    for i in range(1, len(numbers) + 1):
        datas += list(map(''.join, permutations(list(numbers), i)))
    datas = set(datas)

    for data in datas:
        if data[0] == '0':
            continue
        if isPrime(int(data)):
            answer += 1

    return answer

print(solution("17"))
print(solution("011"))

File: programmers/high_score_kit/완전탐색/소수찾기.js
const permutation = (arr, selectNum) => {
    let result = [];
    if (selectNum === 1) return arr.map((v) => [v]);

    arr.forEach((v, idx, arr) => {
        const fixer = v;
        const restArr = arr.filter((_, index) => index !== idx);
        const permuationArr = permutation(restArr, selectNum - 1);
        const combineFixer = permuationArr.map((v) => [fixer, ...v]);
        result.push(...combineFixer);
    });
    return result;
}

const isPrime = (num) => {
    if (num === 0 || num === 1) {
        return false;
    }
    if (num === 2) {
        return true;
    }
    for (let i = 2; i <= Math.floor(Math.sqrt(num)); i++) {
        if (num % i === 0) {
            // 한 번이라도 나누어 졌으니 소수가 아니므로 return false
            return false;
        }
    }
    // 나눠진 수가 없다면 해당 수는 소수이므로 return true
    return true;
}

function solution(numbers) {
    let answer = 0;
    let permuList = [];
    let numberList = [];
    for (let i = 1; i <= numbers.length; i++) {
        permuList.push(...permutation(numbers.split(''), i));
    }
    for (let i = 0; i < permuList.length; i++) {
        numberList.push(Number(permuList[i].join('')));
    }
    numberList = Array.from(new Set(numberList));
    for (let i = 0; i < numberList.length; i++) {
        if (isPrime(numberList[i])) {
            answer += 1;
        }
    }
    return answer;
}

console.log(solution("17")); // 3